### PR TITLE
Product photo

### PR DESCRIPTION
**Pull request by lcegit**
PT Link https://www.pivotaltracker.com/story/show/154132379

### CODE REVIEW

**Review by tochman — DISMISSED** [review]
There are trace of merge conflicts in the code. Please review on GitHub

**Comment by lcegit** [comment]
I had to create a new PR for this feature. All previous comments have been taken in into account.

**Review by tochman — DISMISSED** [review]
I don't mind this for now, but I'm generally against hardcoded solutions like this. I'll approve this but it's up to the team to discuss this solution and make a decision on if this should be merged or not.

**Comment by tochman** [comment]
Conflicts need to be resolved

**Comment by lcegit** [comment]
@NegativeKarma @SophieSjoberg @oliverochman With this solution we're hardcoding the images pipeline -- forcing the admin user to name images exactly as the product's name. Also, each restaurant admin will have to use and access separate images folders. Any final thoughts before we get this merged? Are we ok with this solution for now?

**Comment by tochman** [comment]
I would store an url in an attribute and just point to a resource on the internet for now. We will look into image upload with Paperclip next week.

**Comment by lcegit** [comment]
Ok. So I'll revise the product model and give it a new url attibute?

**Comment by SophieSjoberg** [comment]
@lcegit I can help you with storing url if you want. I don't think we should have it hardcoded.

**Comment by lcegit** [comment]
@SophieSjoberg I added an "image link" column to the product table. Implementing a method to show the image now. Do you think we need to include anything else at this point?

**Comment by lcegit** [comment]
@tochman Would you consider this branch mergeable now after our latest fix?

**Review by tochman — DISMISSED** [review]
Looks okay, I'm a bit worried why you had to update all features with the image link. It should be enough to add the url to the factory.

Also, make sure you test what happens in the view if there is no image. Displaying a broken image is not good UX

**Comment by lcegit** [comment]
@tochman Not sure why it's being required by other feature backgrounds.

Good point, I'll work on that sad path.

File: features/user_can_add_products_to_an_order.feature [hunk 1]
```diff
@@ -8,11 +8,11 @@ Feature: User adds products to an order
...
       | Oliver's Burger   | Kungsgatan 1      | Göteborg |
 
-    Given the following products exist within a specific restaurant and category
-      | name              | restaurant      | category    | menu       | price |
-      | Chicken wings     | Oliver's Burger | Starter     | Lunch      | 45    |
-      | Double Burger     | Oliver's Burger | Main Course | À la carte | 50    |
-
+    Given the following products exist within a specific restaurant
+    | name          | restaurant      | description      | category    | price | image_file_link                    |
+    | Chicken wings | Oliver's Burger | Nice wings       | Starter     | 50 kr | http://www.example.com/image.jpg   |
```
Comment (lcegit) [inline features/user_can_add_products_to_an_order.feature]: Not sure why but the image is required for this test to pass.

File: features/product_list.feature [hunk 2]
```diff
@@ -11,10 +11,10 @@ Feature: User can see product list on the restaurant show page
     | Borat's Palace    | Östra Hmangatan 1 | Göteborg |
 
     Given the following products exist within a specific restaurant
-    | name          | restaurant      | description      | category    | price |
-    | Chicken wings | Oliver's Burger | Nice wings       | Starter     | 50 kr |
-    | Double Burger | Oliver's Burger | Tasty fat burger | Main Course | 50 kr |
-    | Apple pie     | Oliver's Burger | Sweet pie        | Dessert     | 50 kr |
+    | name          | restaurant      | description      | category    | price | image_file_link                    |
+    | Chicken wings | Oliver's Burger | Nice wings       | Starter     | 50 kr | http://www.example.com/image.jpg   |
```
Comment (lcegit) [inline features/product_list.feature]: The image bg column seems to be required for the test to pass.

File: features/product_description.feature [hunk 3]
```diff
@@ -11,10 +11,10 @@ Feature: User can see product description on the restaurant show page
       | Borat's Palace    | Östra Hmangatan 1 | Göteborg |
 
     Given the following products exist within a specific restaurant
-      | name          | restaurant      | description      | price | category    | menu  |
-      | Chicken wings | Oliver's Burger | Nice wings       | 50    | Starter     | Lunch |
-      | Double Burger | Oliver's Burger | Tasty fat burger | 150   | Main Course | Lunch |
-      | Apple pie     | Oliver's Burger | Sweet pie        | 65    | Dessert     | Lunch |
+      | name          | restaurant      | description      | price | category    | menu  | image_file_link                    |
+      | Chicken wings | Oliver's Burger | Nice wings       | 50    | Starter     | Lunch | http://www.example.com/image.jpg   |
```
Comment (lcegit) [inline features/product_description.feature]: The img col is required by the test.

**Comment by lcegit** [comment]
@tochman @SophieSjoberg @oliverochman @NegativeKarma any objections on getting this merged?

**Comment by SophieSjoberg** [comment]
I think it looks good, way to go Lisa! Just wondering if we can add a sad path for broken images or if they are somehow not visible...?

**Comment by lcegit** [comment]
@SophieSjoberg Good point, I don't think it's possible since the string is being read in either case – it isn't being tested for actual validity. Have googled around for solutions but it seems to be out of reach for our scope.

**Review by NegativeKarma — COMMENTED** [review]
Me likey

File: features/product_list.feature (hunk 2 above)
Comment (diraulo) [reply features/product_list.feature]: It should only be required if you want to override the value set in the factory.. which isn't the case for most of those steps.
Your tests should in principle pass if you set the default value in the factory to `http://www.example.com/image.jpg` and override that when needed.

**Review by diraulo — APPROVED** [review]
No objections to this one getting merged

**Review by diraulo — DISMISSED** [review]
Gotta fix the merge conflict first though

**Comment by lcegit** [comment]
@diraulo Sure, however the converted method to show the image doesn't work, have tried multiple configurations for this.

**Comment by diraulo** [comment]
Sure thing. We can probably get this merged once the conflicts are resolved

**Comment by lcegit** [comment]
@diraulo LGTM :)

**Review by SophieSjoberg — DISMISSED** [review]
LGTM 👍

File: features/product_photo.feature [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    Given the following products exist within a specific restaurant
+    | name          | restaurant      | description      | category    | price | image_file_link                    |
+    | Chicken wings | Oliver's Burger | Nice wings       | Starter     | 50 kr | http://www.example.com/image.jpg   |
+    | Double Burger | Oliver's Burger | Tastiest burger  | Main        | 50 kr |                                    |
+
+  @googlemap
+  Scenario:
+    When I visit the "Oliver's Burger" show page
+    Then I would like to see "Nice wings" under "Chicken wings" with a price of "50.0 kr"
+    Then show me the page
```
Comment (oliverochman) [inline features/product_photo.feature]: maybe remove this

**Review by diraulo — APPROVED** [review]
:shipit:

**Pull request merged**